1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+
+```powershell
+false || Write-Output "Reached"
+```
+
+```output
+Reached
+```
+
+```powershell
+false || Write-Output "Command failed" && Write-Output "Backup"
+```
+
+```output
+Command failed
+Backup
+```
+
+Also see: [intended test cases for implementation](https://github.com/PowerShell/PowerShell/blob/f8b899e42c86957a1b58273be0029f40a67bf1b6/test/powershell/Language/Operators/BashControlOperator.Tests.ps1).
+
+## Specification
+
+### Grammar
+
+Pipeline chains will be implemented using the already reserved `&&` and `||` operators.
+These operators will have the following grammar:
+
+```none
+statement:
+    | ... <existing statements except pipeline>   # Other statements
+    | variable_expression "=" statement           # Assignment
+    | pipeline_chain ["&"]                        # Pipeline chains
+
+pipeline_chain:
+    | pipeline
+    | pipeline_chain "&&" [newlines] pipeline
+    | pipeline_chain "||" [newlines] pipeline
+```
+
+#### Optional pipeline chaining
+
+A pipeline chain is a chain of one or more pipelines
+and takes the place of a pipeline in the current PowerShell syntax.
+
+In the degenerate case, a pipeline chain of a single pipeline
+will work exactly as pipelines currently do in PowerShell,
+so there is no breakage to the existing grammar.
+
+#### Line continuation
+
+After a pipeline chain operator,
+any newlines will be skipped in anticipation of the following pipeline.
+
+For example, the following would be a single pipeline chain:
+
+```powershell
+cmd1 &&
+    cmd2 ||
+    cmd3
+```
+
+If the end of file is reached after a pipeline chain operator,
+incomplete input will be reported so that integrating tools
+will know to keep prompting for input.
+
+#### Left-associativity
+
+Like in POSIX shells, pipeline chain operators will be left associative,
+meaning chain operators will group from left to right.
+
+For example, given the following pipeline chain:
+
+```powershell
+cmd1 && cmd2 || cmd3
+```
+
+This will be grouped as:
+
+```none
+[[cmd1 && cmd2] || cmd3]
+```
+
+As a syntax tree, this would look like:
+
+```none
+            "||"
+          /      \
+        "&&"     "cmd3"
+      /      \
+   "cmd1"    "cmd2"
+```
+
+With the syntax tree deepening on the left as more operators are chained.
+
+Semantically, this means that `cmd1 && cmd2` would be evaluated first,
+and its result used to govern the evaluation of `|| cmd3`.
+
+#### Higher precedence than `&` and `;`
+
+Pipeline chain operators will have higher precedence than
+pipeline background operators (`&`)
+or statement separators (`;`).
+
+For example:
+
+```none
+cmd1 && cmd2 &
+```
+
+Will bind as:
+
+```none
+[[cmd1 && cmd2] &]
+```
+
+Having the syntax tree:
+
+```none
+            <statement>
+           /           \
+         "&&"          "&"
+        /    \
+     "cmd1"  "cmd2"
+```
+
+The consequence of this will be that an entire pipeline chain
+can be sent to a background job for evaluation,
+rather than individual pipelines within it.
+

```

Comment:
It is worth calling out right here that this is the opposite of bash. I also question whether or not this is the right thing to do for that reason. I know you have additional comments below, but if this is really the way we want to move forward here, attention needs to be drawn to this fact, coupled with reasoning why this is better. In the past though, the Linux community has not responded favorably to differing behaviours with commands, and I worry they will respond similarly if we change operator precedence for operators. Does it really buy us enough to change the precedence, or should we just follow bash?

Reply:
I believe this proposal makes more sense.

Reply:
> Does it really buy us enough to change the precedence, or should we just follow bash?

Actually we already changed the precedence. I think I mention it under **Background operator changes** (line 446 currently).

In PowerShell, the background operator applies to a pipeline rather than a statement.

So currently in a return statement, the background operator binds to the pipeline subordinate to the statement. This function returns a job:

```powershell
function Test
{
   return Write-Output "something" &
}
```

Similarly with `throw`:

```powershell
function Bad
{
    throw "Message" &
}
```

The function `Bad` does not succeed, it throws a Job. If the background operator applied to the statement as a whole (wherein the statement is backgrounded) the `throw` itself would be backgrounded and the function would succeed.

Also consider:

```powershell
foreach ($x in 1..100) { break & }
```

Reply:
I personally think that elevating backgrounding to statement level invites serious inconsistencies and complexities into the grammar, but I thought it was worth writing up since I discussed it with @daxian-dbw.